pkg/webhook/conversion/testData/pkg/apis/jobs/v1/externaljob_types.go

```diff
@@ -0,0 +1,97 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/conversion/testData/pkg/apis/jobs/v2"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// ExternalJobSpec defines the desired state of ExternalJob
+type ExternalJobSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+	RunAt string `json:"runAt"`
+}
+
+// ExternalJobStatus defines the observed state of ExternalJob
+type ExternalJobStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ExternalJob is the Schema for the externaljobs API
+// +k8s:openapi-gen=true
+type ExternalJob struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ExternalJobSpec   `json:"spec,omitempty"`
+	Status ExternalJobStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ExternalJobList contains a list of ExternalJob
+type ExternalJobList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ExternalJob `json:"items"`
+}
+
+func init() {
+	SchemeBuilder.Register(&ExternalJob{}, &ExternalJobList{})
+}
+
+// since v1.ExternalJob is a spoke, it needs to be convertable.
+// It needs to implement convert to/from storage version
+// TODO(droot): evaluate advantages of taking `conversion.Hub` as input to
```

Comment:
>Is there a good way to remove one of the two?
Annotations (we have started calling them markers now because annotation/tags are so overloaded in k8s world :)) is primarily needed for controller-gen to generate the manifests for conversion webhook. Only way to remove them would be to make controller-gen parse these types and examine if they implement these methods (which can be done, but is a bit stretch looking at current state of the code in controller-gen). Till then, this is the only option.

>I had kinda expected ConvertTo and ConvertFrom to take a runtime.Object initially but I concluded that it's probably a disadvantage as you lose some compile time checks theoretically
Agreed. using `Hub` has type-safety as well readability from an end-user's perspective. So will took out that TODO.